fedimint-server/src/multiplexed.rs

```diff
@@ -66,20 +59,99 @@ struct ModuleMultiplexerInner<MuxKey, Msg> {
 /// This type is thread-safe and can be cheaply cloned.
 #[derive(Clone)]
 pub struct PeerConnectionMultiplexer<MuxKey, Msg> {
-    inner: Arc<ModuleMultiplexerInner<MuxKey, Msg>>,
+    /// Sender of send requests
+    send_requests_tx: Sender<(Vec<PeerId>, MuxKey, Msg)>,
+    /// Sender of receive callbacks
+    receive_callbacks_tx: Sender<Callback<MuxKey, Msg>>,
+    /// Sender of peer bans
+    peer_bans_tx: Sender<PeerId>,
 }
 
+type Callback<MuxKey, Msg> = (MuxKey, oneshot::Sender<(PeerId, Msg)>);
+
 impl<MuxKey, Msg> PeerConnectionMultiplexer<MuxKey, Msg>
 where
-    Msg: Serialize + DeserializeOwned + Unpin + Send + Debug,
-    MuxKey: Serialize + DeserializeOwned + Unpin + Send + Debug + Eq + Hash + Clone,
+    Msg: Serialize + DeserializeOwned + Unpin + Send + Debug + 'static,
+    MuxKey: Serialize + DeserializeOwned + Unpin + Send + Debug + Eq + Hash + Clone + 'static,
 {
     pub fn new(connections: PeerConnections<ModuleMultiplexed<MuxKey, Msg>>) -> Self {
+        let (send_requests_tx, send_requests_rx) = channel(1000);
+        let (receive_callbacks_tx, receive_callbacks_rx) = channel(1000);
+        let (peer_bans_tx, peer_bans_rx) = channel(1000);
+
+        tokio::spawn(Self::run(
+            connections,
+            Default::default(),
+            send_requests_rx,
+            receive_callbacks_rx,
+            peer_bans_rx,
+        ));
+
         Self {
-            inner: Arc::new(ModuleMultiplexerInner {
-                connections: Mutex::new(connections),
-                out_of_order: Default::default(),
-            }),
+            send_requests_tx,
+            receive_callbacks_tx,
+            peer_bans_tx,
+        }
+    }
+
+    async fn run(
+        mut connections: PeerConnections<ModuleMultiplexed<MuxKey, Msg>>,
+        mut out_of_order: ModuleMultiplexerOutOfOrder<MuxKey, Msg>,
+        mut send_requests_rx: Receiver<(Vec<PeerId>, MuxKey, Msg)>,
+        mut receive_callbacks_rx: Receiver<Callback<MuxKey, Msg>>,
+        mut peer_bans_rx: Receiver<PeerId>,
+    ) -> Cancellable<()> {
+        loop {
+            let mut key_inserted: Option<MuxKey> = None;
+            tokio::select! {
+                 // Send requests are forwarded to underlying connections
+                 send_request = send_requests_rx.recv() => {
```

Comment:
User sends the send request over a channel, just so this thread receives it, and then turns over and calls `connections.send` which under the hood puts it on another channel..